nni/mutable/frozen.py

```diff
@@ -0,0 +1,226 @@
+    if sample is None:
+        ctx = frozen_context.current()
+        if ctx is not None:
+            sample = ctx
+
+    if sample is not None:
+        # If we have a sample, we can just use it.
+        # Use freeze here to detect potential label mismatch errors.
+        try:
+            return mutable.freeze(sample)
+        except:
+            _logger.error(
+                'Failed to freeze mutable %s with sample %s. '
+                'In NAS, please make sure to have registered it via add_mutable(). '
+                'Otherwise, please make sure you are not inside a frozen_context.',
+                mutable, sample)
+            raise
+    else:
+        if retries < 0:
+            raise RuntimeError('Cannot freeze mutable. Please provide a context.')
+
+        _logger.warning('No frozen context is found for %s. Assuming no context.', repr(mutable))
+        return mutable.robust_default(retries=retries)
+
+
+class frozen_context(ContextStack):
+    """
+    Context manager to set a sample into context.
+    Then the sample will be retrievable from an arbitrary level of function calls via :func:`current_frozen_context`.
+
+    There are two use cases:
+
+    1. Setting a global sample so that some modules can directly create the frozen version, rather than first-create-and-freeze.
+    2. Sharing default / dry-run samples when the search space is dynamically created.
+
+    The implementation is basically adding another layer of empty dict on top of a global stack.
+    When retrieved, all dicts in the stack will be merged, from the bottom to the top.
+    When updated, only the dict on the top will be updated.
+
+    Parameters
+    ----------
+    sample
+        The sample to be set into context.
+
+    Returns
+    -------
+    ContextStack
+        Context manager that provides a frozen context.
+
+    Examples
+    --------
+    ::
+        def some_func():
+            print(frozen_context.current()['learning_rate'])  # 0.1
+
+        with frozen_context({'learning_rate': 0.1}):
+            some_func()
+    """
+
+    def __init__(self, sample: Sample | None = None):
+        super().__init__(_FROZEN_CONTEXT_KEY, sample or {})
+
+    @staticmethod
+    def top_context() -> frozen_context:
+        return ContextStack.top(_FROZEN_CONTEXT_KEY)
+
+    @staticmethod
+    def current() -> dict | None:
+        """Retrieve the current frozen context.
+        If multiple layers have been found, they would be merged from bottom to top.
+
+        Returns
+        -------
+        The sample in frozen context.
+        If no sample is found, return none.
+        """
+        try:
+            ContextStack.top(_FROZEN_CONTEXT_KEY)
+            sample: Sample = {}
+            for ctx in ContextStack.stack(_FROZEN_CONTEXT_KEY):
+                if not isinstance(ctx, dict):
+                    raise TypeError(f'Expect architecture to be a dict, found: {ctx}')
+                sample.update(ctx)
```

Comment:
just curious, key in `ctx` will not conflict?

Reply:
It will. The top of the stack will overwrite the bottom.